File: ros2_ws/ser_cli/src/cpp_order_system/src/kitchen_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
#include <memory>

// 1. 實作 take_order 回呼函式，參數為 Request 與 Response 的 shared_ptr
void take_order(const std::shared_ptr<example_interfaces::srv::AddTwoInts::Request> request,
          std::shared_ptr<example_interfaces::srv::AddTwoInts::Response> response)
{
    // 2. 計算 request 內的 a(漢堡) + b(薯條) 並填入 response->sum
    response->sum = request->a + request->b;
    // 3. 印出收到的訂單內容 (RCLCPP_INFO)
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "收到訂單：漢堡 %ld 份, 薯條 %ld 份", request->a, request->b);
}

int main(int argc, char **argv)
{
  // 4. 初始化 rclcpp
  rclcpp::init(argc, argv);
  // 5. 建立節點 "kitchen_node"
  auto node = rclcpp::Node::make_shared("kitchen_node");
  // 6. 建立服務端 (create_service)，名稱取為 "place_order"，並綁定上述的 take_order 函式
  auto service = node->create_service<example_interfaces::srv::AddTwoInts>("place_order", take_order);
  // 7. 啟動節點 (spin)、關閉 rclcpp
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/ser_cli/src/py_order_system/py_order_system/kitchen_server.py
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts # 借用內建格式：傳入 a, b，回傳 sum

class KitchenServer(Node):
    def __init__(self):
        super().__init__('kitchen_node')
        # 建立服務伺服器：(服務型態, '服務名稱', 回呼函式)
        self.srv = self.create_service(AddTwoInts, 'place_order', self.order_callback)
        self.get_logger().info('👨‍🍳 廚房已開工，等待點餐...')

    def order_callback(self, request, response):
        # 讀取 Client 傳來的 request (a=漢堡數量, b=薯條數量)
        burgers = request.a
        fries = request.b
        
        # 計算總數後賦值給 response.sum
        response.sum = burgers + fries
        
        # 在終端機印出收到什麼訂單
        self.get_logger().info(f'🔔 收到新訂單: 漢堡 {burgers} 份, 薯條 {fries} 份')
        
        return response # 必須將 response 物件回傳給 Client

def main():
    rclpy.init()
    node = KitchenServer()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()